File: golang/LinkedList/singlelist.go
package main

import (
	"fmt"
)

//set node struct
type Node struct {
	data int
	next *Node
}

//set list struct
type LinkedList struct {
	head *Node
}

//一般的list.head 就是整个LinkedList
//Head Insert

//        -----------
//|head|->|data|next| ->
//		  -----------
//如果链表为空，则直接把data变量赋值给list.head
//如果链表非空，则把list.head的值，赋值给data.next，因为在头部插入，插入的数据就变为头部数据了
func (list *LinkedList) InsertFirst(i int) {
	data := &Node{data: i}
	if list.head != nil {
		data.next = list.head
	}
	list.head = data
	return
}

//如果链表为空，则直接把data变量赋值给list.head
//如果链表非空，则for循环找到最后一个节点，最后一个节点的next指向data
func (list *LinkedList) InsertLast(i int) {
	data := &Node{data: i}
	if list.head == nil {
		list.head = data
		return
	}
	current := list.head
	for current.next != nil {
		current = current.next
	}
	current.next = data
}

//在指定节点index前插入
//如果链表插入的节点小于等于1，使用InsertFirst
//如果链表插入的节点大于链表的长度，使用InsertLast
//golang 中的for 和 while
func (list *LinkedList) Insert(index int, i int) {
	if index <= 0 {
		list.InsertFirst(i)
	} else if index > list.GetSize() {
		list.InsertLast(i)
	} else {
		pre := list.head
		//链表下标从0开始
		//这样可以从1开始
		//count := 1
		//for count < (index - 1) {
		//	pre = pre.next
		//	count++
		//}
		count := 1
		for count < index {
			pre = pre.next
			count++
		}
		node := &Node{data: i}
		node.next = pre.next
		pre.next = node

	}
}

func (list *LinkedList) RemoveByIndex(index int) bool {
	if list.head == nil {
		return false
	}
	if index < 0 {
		return false
	} else if index == 0 {
		list.head = list.head.next
		return true
	}
	current := list.head
	count := 1
	for count < index && current.next != nil {
		current = current.next
		count++
	}
	current.next = current.next.next
	return true
}

func (list *LinkedList) RemoveByValue(data int) bool {
	if list.head == nil {
		return false
	}
	if list.head.data == data {
		list.head = list.head.next
		return true
	}
	current := list.head
	for current.next != nil {
		if current.next.data == data {
			current.next = current.next.next
			return true
		}
		current = current.next
	}
	return true
}

func (list *LinkedList) GetSize() int {
	count := 0
	current := list.head
	for current != nil {
		count += 1
		current = current.next
	}
	return count
}

func (list *LinkedList) GetItems() []int {
	var items []int
	current := list.head
	for current != nil {
		items = append(items, current.data)
		current = current.next
	}
	return items
}

func (list *LinkedList) SearchValue(i int) bool {
	if list.head == nil {
		return false
	}
	current := list.head
	for current.next != nil {
		if current.next.data == i {
			return true
		}
		current = current.next
	}
	return false
}

func (list *LinkedList) GetFirst() (int, bool) {
	if list.head == nil {
		return 0, false
	}
	return list.head.data, true
}

func (list *LinkedList) GetLast() (int, bool) {
	if list.head == nil {
		return 0, false
	}
	current := list.head
	for current.next != nil {
		current = current.next
	}
	return current.data, true
}

func main() {
	list := &LinkedList{}
	list.InsertFirst(1)
	list.InsertFirst(3)
	list.InsertFirst(5)
	list.InsertFirst(7)
	list.InsertFirst(9)
	list.InsertLast(2)
	fmt.Println(list.GetSize())
	fmt.Println(list.GetItems())
	list.Insert(1, 10)
	//list.Insert(0, 11)
	list.RemoveByIndex(4)
	fmt.Println(list.GetSize())
	fmt.Println(list.GetItems())
	list.RemoveByValue(1)
	list.RemoveByValue(2)
	fmt.Println(list.GetSize())
	fmt.Println(list.GetItems())
	fmt.Println(list.SearchValue(5))
	fmt.Println(list.GetFirst())
	fmt.Println(list.GetLast())
}

File: python/LinkedList/SingleList.py
#!/user/bin/python3

class Node(object):
    def __init__(self, data):
        self.data = data
        self.next = None
    
class SingleList(object):
    
    def __init__(self, node=None):
        self.__head = node
    
    def is_empty(self):
        return self.__head == None
    
    def insert_first(self, value):
        node = Node(value)
        node.next = self.__head
        self.__head = node
    
    def insert_last(self, value):
        node = Node(value)
        current = self.__head
        if self.__head == None:
            self.__head = node
        else:
            while current.next:
                current = current.next
            current.next = node
    
    def insert(self, index, value):
        if index < 0:
            self.add(value)
        elif index > self.get_len():
            self.append(value)
        else:
            current = self.__head
            count = 1
            while count < index:
                current = current.next
                count += 1

            node = Node(value)
            node.next = current.next
            current.next = node

    def get_len(self):
        if self.is_empty():
            return False
        else:
            count = 0
            current = self.__head
            while current != None:
                current = current.next
                count += 1
            return count

    def delete_byvalue(self, value):
        if self.is_empty():
            return False
        else:
            current = self.__head
            while current.next != None:
                if current.next.data == value:
                    current.next = current.next.next
                    return True
                else:
                    current = current.next
            return True
    
    def delete_byindex(self, index):
        if self.is_empty():
            return False
        elif index < 0 or index > self.get_len():
            return False
        elif index == 0:
            current = self.__head
            self.__head = current.next
        else:
            current = self.__head
            count = 1
            while count < index:
                current = current.next
                count += 1
            current.next = current.next.next
            return True
    
    def search_by_value(self, value):
        if self.is_empty():
            return False
        current = self.__head
        while current.next != None:
            if current.next.data == value:
                return value
            current = current.next
        return False

    def search_by_index(self, index):
        if self.is_empty():
            return False
        if index < 0 or index > self.get_len():
            return False
        elif index == 0:
            return self.__head.data
        else:
            count = 1
            current = self.__head
            while count < index:
                current = current.next
                count += 1
            return current.next.data

    def get_items(self):
        if self.is_empty():
            return False
        new_list = []
        current = self.__head
        while current != None:
            new_list.append(current.data)
            current = current.next
        return new_list

if __name__ == '__main__':
    singlelist = SingleList()
    print(singlelist.is_empty())
    print(singlelist.get_len())
    singlelist.insert_first(1)
    print(singlelist.is_empty())
    print(singlelist.get_len())
    singlelist.insert_last(2)
    singlelist.insert_last(3)
    singlelist.insert_last(5)
    singlelist.insert(2, 10)
    print(singlelist.get_len())
    print(singlelist.get_items())
    print(singlelist.search_by_value(2))
    print(singlelist.search_by_index(2))
    singlelist.delete_byindex(0)
    print(singlelist.get_len())
    print(singlelist.get_items())


        

